ValidModeSetupHandler.lua

```diff
@@ -0,0 +1,144 @@
+			end
+			if #disallowedSetups > 0 then 
+				validData.disallowedSetups =  disallowedSetups
+			end
+			modeSetup[i+1] = validData
+			i = i+1
+		end
+		self.modeSetup = modeSetup
+	--	DebugUtil.printTableRecursively(modeSetup, "   ", 0, 3)
+	else 
+		print(string.format("ValidModeSetupHandler, could not load file: %s",filePath))
+	end
+end
+
+--load the Setup for either the allowed or disallowed Specs
+function ValidModeSetupHandler:loadSetupsFromXml(xmlFile,baseKey,entry,allowKey)
+	local key = string.format("%s.%s",baseKey,allowKey)
+	local i = 0
+	while true do 
+		local setupKey = string.format("%s.%s(%d)",key,"Setup",i)
+		if not hasXMLProperty(xmlFile, setupKey) then
+		--	courseplay.info(string.format("ValidModeSetupHandler, %s not found: %s",allowKey,setupKey))
+			break
+		end
+--		print(string.format("ValidModeSetupHandler, %s found",setupKey))
+		local specSetup = {}
+		self:loadSetupFromXml(xmlFile,setupKey,specSetup,"Specialization")	
+		self:loadSetupFromXml(xmlFile,setupKey,specSetup,"SpezialTool")	
+		if #specSetup>0 then
+			table.insert(entry,specSetup)
+		end
+		i = i+1
+	end
+end
+--SpezialTool
+--load the Specializations for either the allowed or disallowed Specs
+function ValidModeSetupHandler:loadSetupFromXml(xmlFile,baseKey,entry,itemKey)
+	local i = 0
+	while true do 
+		local key = string.format("%s.%s(%d)",baseKey,itemKey,i)
+		if not hasXMLProperty(xmlFile, key) then
+		--	courseplay.info(string.format("ValidModeSetupHandler, %s not found: %s",itemKey,key))
+			break
+		end
+		local specName = getXMLString(xmlFile, key.."#name")
+	--	print(string.format("ValidModeSetupHandler, %s found: %s at: %s",itemKey,specName,key))
+		table.insert(entry,specName)
+		i = i+1
+	end
+end
+
+
+---checks if the "mode" is allowed and not disallowed for the vehicle
+---@param int : cpMode to check
+---@param vehicle : object, implement,vehicle
+---@return boolean : isAllowedOkay, boolean : isDisallowedOkay
+function ValidModeSetupHandler:isModeValid(mode,object)
+	local validData = self.modeSetup[mode]
+	local isAllowedOkay = false
+	local isDisallowedOkay = true
+	if validData then 
+		if validData.allowedSetups then 
+			isAllowedOkay = self:isSetupAllowedValid(validData.allowedSetups,object)
+			courseplay.debugVehicle(18,object,"allowedSetups, mode: %d, isAllowedOkay: %s",mode,tostring(isAllowedOkay))
+		end
+		if validData.disallowedSetups then 
+			isDisallowedOkay = self:isSetupDisallowedValid(validData.disallowedSetups,object)
+			courseplay.debugVehicle(18,object,"disallowedSetup, mode: %d,, isDisallowedOkay: %s",mode,tostring(isDisallowedOkay))
+		end
+	else 
+		courseplay.info("ValidModeSetupHandler, validData==nil !!")
+		return false
+	end
+	return isAllowedOkay, isDisallowedOkay
+end
+
+---checks if one setup combo is allowed for this object
+---@param table : setups 
+---@param vehicle : object, implement,vehicle
+---@return boolean : isAllowedOkay
+function ValidModeSetupHandler:isSetupAllowedValid(setups,object)
```

Comment:
Wondering if it was possible to use the standard Giants SpecializationUtil.hasSpecialization()? Also using the specialization class names in the XML instead of spec_something?

Reply:
That would be a great Idee, but I am not sure how to translate the XML String name back to a class to use  SpecializationUtil.hasSpecialization() and compare the classes.